hera_cal/frf.py

```diff
@@ -642,30 +1008,110 @@ def tophat_frfilter(self, keys=None, wgts=None, mode='clean',
-            center_frates = {k: 0.0 for k in keys}
+        if uvb is None and max_frate_coeffs is None:
+            # if max_frate_coeffs is none and uvb is none, fringe-rate filter all modes that could be occupied by sky emission.
+            frate_centers, frate_half_widths = sky_frates(self.hd, keys=keys, frate_standoff=frate_standoff,
+                                                          frate_width_multiplier=frate_width_multiplier, min_frate_half_width=min_frate_half_width)
+        elif uvb is None and max_frate_coeffs is not None:
+            # if uvb is None and max_frate_coeffs is not None, use max_frate_coeffs.
+            frate_half_widths = {k: np.max([max_frate_coeffs[0] * self.blvecs[k[:2]][0] + max_frate_coeffs[1], 0.0]) for k in keys}
+            frate_centers = {k: 0.0 for k in keys}
+
+        elif uvb is not None or frate_profiles is not None:
+            # if uvb is not None, get fringe-rates from binning.
+            frate_centers, frate_half_widths = get_fringe_rate_limits(self.hd, uvb, nfr=nfr, dfr=dfr,
+                                                                      percentile_low=percentile_low,
+                                                                      percentile_high=percentile_high,
+                                                                      keys=keys, verbose=verbose,
+                                                                      frate_standoff=frate_standoff, fr_freq_skip=fr_freq_skip,
+                                                                      frate_width_multiplier=frate_width_multiplier,
+                                                                      min_frate_half_width=min_frate_half_width,
+                                                                      frate_profiles=frate_profiles)
+
         wgts = io.DataContainer({k: (~self.flags[k]).astype(float) for k in self.flags})
+        if pre_filter_modes_between_lobe_minimum_and_zero:
+            self.pre_filter_resid = DataContainer({})
+            self.pre_filter_resid_flags = DataContainer({})
+            filter_kwargs_no_data = copy.deepcopy(filter_kwargs)
+            if 'data' in filter_kwargs_no_data:
+                del filter_kwargs_no_data['data']
         for k in keys:
             if mode != 'clean':
                 filter_kwargs['suppression_factors'] = [tol]
             else:
                 filter_kwargs['tol'] = tol
-            self.fourier_filter(keys=[k], filter_centers=[center_frates[k]], filter_half_widths=[width_frates[k]],
-                                mode=mode, x=self.times * 3.6 * 24.,
-                                data=self.data, flags=self.flags, wgts=wgts,
-                                ax='time', cache=filter_cache, skip_wgt=skip_wgt, verbose=verbose, **filter_kwargs)
+
+            if pre_filter_modes_between_lobe_minimum_and_zero:
```

Comment:
this seems like something that should really be dealt with as a separate call to this function, or is there some benefit (other than convenience) that this offers?

Reply:
I think it would make sense if the fringe rate limit containers were computed outside of this function. Perhaps this is another reason to build another layer of driver methods.